File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/EventNotificationSample/src/com/hp/workpath/sample/eventnotificationsample/handler/StandaloneResultHandler.java
package com.hp.workpath.sample.eventnotificationsample.handler;

import android.content.Context;
import android.util.Log;

import com.hp.workpath.sample.eventnotificationsample.task.InitializationTask;
import com.hp.workpath.sample.eventnotificationsample.task.ResultHandler;
import com.hp.workpath.sample.eventnotificationsample.task.UserDetailsReaderTask;
import com.hp.workpath.sample.eventnotificationsample.util.PreferenceManager;

/**
 * A standalone implementation of ResultHandler that is not tied to the BroadcastReceiver lifecycle.
 * This handler will continue to work even after the BroadcastReceiver has been destroyed.
 */
public class StandaloneResultHandler implements ResultHandler {
    public static final String TAG = "[SAMPLE]" + "StandaloneHandler";
    
    private final Context mAppContext;  // Use application context to avoid memory leaks
    private final String mEventAction;
    private InitializationTask mInitTask;
    
    public StandaloneResultHandler(Context context, String eventAction) {
        // Store application context to avoid memory leaks
        this.mAppContext = context.getApplicationContext();
        this.mEventAction = eventAction;
    }
    
    public void setInitTask(InitializationTask task) {
        this.mInitTask = task;
    }
    
    @Override
    public void handleComplete() {
        try {
            new UserDetailsReaderTask(mAppContext, this).taskExecute();
        } catch (Exception e) {
            Log.e(TAG, "Exception in handleComplete: " + e.getMessage());
            cleanup();
        }
    }
    
    @Override
    public void handleException(Throwable t) {
        Log.e(TAG, "handleException: " + t.getMessage());
        cleanup();
    }
    
    @Override
    public void handleUpdate(String currentPrincipal) {
        try {
            Log.i(TAG, "handleUpdate: " + currentPrincipal);
            PreferenceManager.saveEvent(mAppContext, mEventAction, currentPrincipal);
        } catch (Exception e) {
            Log.e(TAG, "Exception in handleUpdate: " + e.getMessage());
        } finally {
            cleanup();
        }
    }
    
    /**
     * Clean up resources when all tasks are finished
     */
    private void cleanup() {
        if (mInitTask != null) {
            mInitTask.cancel();
            mInitTask = null;
        }
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/EventNotificationSample/src/com/hp/workpath/sample/eventnotificationsample/handler/StandaloneResultHandler.kt
package com.hp.workpath.sample.eventnotificationsample.handler

import android.content.Context
import android.util.Log
import com.hp.workpath.sample.eventnotificationsample.task.InitializationTask
import com.hp.workpath.sample.eventnotificationsample.task.ResultHandler
import com.hp.workpath.sample.eventnotificationsample.task.UserDetailsReaderTask
import com.hp.workpath.sample.eventnotificationsample.util.PreferenceManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * A standalone implementation of ResultHandler that is not tied to the BroadcastReceiver lifecycle.
 * This handler will continue to work even after the BroadcastReceiver has been destroyed.
 */
class StandaloneResultHandler(context: Context, private val eventAction: String) : ResultHandler {
    companion object {
        const val TAG = "[SAMPLE]StandaloneHandler"
    }
    
    // Use application context to avoid memory leaks
    private val appContext: Context = context.applicationContext
    private var initTask: InitializationTask? = null
    
    fun setInitTask(task: InitializationTask) {
        this.initTask = task
    }
    
    override fun handleComplete() {
        try {
            CoroutineScope(Dispatchers.IO).launch {
                val task = UserDetailsReaderTask(appContext, this@StandaloneResultHandler)
                task.taskExecute()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Exception in handleComplete: ${e.message}")
            cleanup()
        }
    }
    
    override fun handleException(t: Throwable?) {
        Log.e(TAG, "handleException: ${t?.message}")
        cleanup()
    }
    
    override fun handleUpdate(currentPrincipal: String?) {
        try {
            Log.i(TAG, "handleUpdate: $currentPrincipal")
            PreferenceManager.saveEvent(appContext, eventAction, currentPrincipal)
        } catch (e: Exception) {
            Log.e(TAG, "Exception in handleUpdate: ${e.message}")
        } finally {
            cleanup()
        }
    }
      /**
     * Clean up resources when all tasks are finished
     */
    private fun cleanup() {
        // Release the reference to the task
        initTask = null
    }
}
